src/smt/smt_engine_scope.cpp

```diff
@@ -46,9 +46,16 @@ ProofManager* currentProofManager() {
 #endif /* IS_PROOFS_BUILD */
 }
 
+ResourceManager* currentResourceManager()
+{
+  return s_smtEngine_current->getResourceManager();
+}
+
 SmtScope::SmtScope(const SmtEngine* smt)
     : NodeManagerScope(smt->d_nodeManager),
-      d_oldSmtEngine(s_smtEngine_current) {
+      d_oldSmtEngine(s_smtEngine_current),
+      d_optionsScope(smt ? &const_cast<SmtEngine*>(smt)->getOptions() : nullptr)
```

Comment:
We should not use `const_cast`. Why not make the argument to SmtScope non-const?

Reply:
Because we have to support making SmtScope on constant SmtEngine methods (where `this` is const).  I had played around with several ways of refactoring this, but for now I am copying the same hack as 2 lines below (line 60) where const_cast is already used.

Reply:
I see. That's really ugly but I suppose there is not much to do and it will be resolved once we get rid of singletons.